[air] add tuner user guide (#26837)
ray-project/ray: doc/source/ray-air/tuner.rst

.. _air-tuner:

Hyperparameter Tuning
=====================

> The Ray AIR Tuner API is the recommended way to launch hyperparameter tuning jobs with Ray Tune.

Reviewer: I think we should phrase this differently. In here it sounds like Ray AIR Tuner is a wrapper over Ray Tune, but actually they should be one and the same. 

How about saying "The Tuner API is the recommended of tuning hyperparameters in Ray AIR."?

Suppose that you already have a Ray AIR trainer at hand.
You have fitted it and obtained a model and training results.

Next, you may want to see if you can change some hyperparameters to further improve the model.

The Tuner API allows you to plug in your existing Trainer. Additionally you can specify the search space
and search algorithms for your hyperparameter tuning job.

In this guide, we will show you how to do this.

.. note::
   The Tuner API can also be used to launch hyperparameter tuning without using Ray AIR Trainers - see
   :ref:`the Ray Tune documentation <tune-main>` for more guides examples.

Basic usage
-----------

> Assuming you already have your own Trainer, this is how the Ray Tune API works in a nutshell:

Reviewer: ```suggestion
Suppose that you already have a Ray AIR trainer at hand. You have fitted it and obtained a model and training results.

Below, we demonstrate how you can plug in your existing Trainer into a Tuner.
```

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __basic_start__
    :end-before: __basic_end__

Defining the search space
-------------------------

> The Tuner API takes in a `param_space` argument where you can define the search space that Ray Tune will

Reviewer: I'm thinking we should not explicitly reference Ray Tune as a separate entity

> The Tuner API takes in a `param_space` argument where you can define the search space that Ray Tune will
> use to try out hyperparameter configurations.

Reviewer: ```suggestion
`Tuner` takes in a `param_space` argument where you can define the search space from which hyperparameter configurations will be sampled.
```

Generally, you can search over most arguments and configurations provided by Ray AIR (and your own options).
This includes Ray Datasets, Preprocessors, the distributed training configuration (ScalingConfig)
and general hyperparameters.

The main exclusion here are all arguments of the :class:`ray.air.config.RunConfig`, which are inherently un-tunable.
Another exclusion is the :class:`ray.tune.tune_config.TuneConfig`, which defines the tuning behavior itself - the settings

> apply to all trials and thus can't be tuned.

Reviewer: Let's move this to the bottom of this subsection.

Examples for common parameters (depending on the Trainer and model) you can tune are:

- The training batch size
- The learning rate for SGD-based training (e.g. image classification)
- The maximum depth for tree-based models (e.g. XGBoost)

The following shows some example code on how to specify the ``param_space``.

.. tabbed:: XGBoost

    .. literalinclude:: doc_code/tuner.py
        :language: python
        :start-after: __xgboost_start__
        :end-before: __xgboost_end__

.. tabbed:: Deep Learning

    .. literalinclude:: doc_code/tuner.py
        :language: python
        :start-after: __torch_start__
        :end-before: __torch_end__

As you can see in the above example, the Tuner API allows you to specify a search space over most parameters:

> - Parameters will be passed to the Trainer

Reviewer: A couple gotchas about the behavior of parameters:

- Dictionaries will be merged
- Scaling configs will be merged
- Duplicate items (which are both defined in the Trainer and Tuner) will be overwritten by the ones defined in the Tuner

The Tuner API also offers the possibility to apply different data preprocessing steps, as shown in the following snippet.

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __tune_preprocess_start__
    :end-before: __tune_preprocess_end__

The Tuner API allows you to even tune the train/validation datasets:

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __tune_dataset_start__
    :end-before: __tune_dataset_end__

In general, all the arguments accepted by your :ref:`Trainer <air-trainer-ref>` can be tuned
(except for the :class:`ray.air.config.RunConfig`).

Specify TuneConfig
------------------
This config contains tuning specific settings, including the tuning algorithm to use, the metric and mode to rank results etc.

> See the :class:`Tuner <ray.tune.tuner.Tuner>` and the :class:`TuneConfig API reference <ray.tune.tune_config.TuneConfig>` for more details.

Reviewer: Add a basic example here?

Specify RunConfig
-----------------
This config contains framework's runtime configurations that are more generic than tuning specific settings.
This may include failure/retry configurations, verbosity levels, the name of the experiment, its logging directory,
checkpoint configuration as well as its syncing configuration.

> See the :class:`RunConfig API reference <ray.air.config.RunConfig>` for more details.

Reviewer: Add a basic example here?

Putting everything together
---------------------------
We can now construct a Tuner and call ``Tuner.fit`` on it!

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __tuner_start__
    :end-before: __tuner_end__

.. note::
    ``num_samples=2`` here will be applied to the whole suite of grid search. In other words,
    we will generate 4 trials.

> For a more end-to-end example, checkout our :doc:`/ray-air/examples/analyze_tuning_results` guide.

Reviewer: This guide only for analysis right?

Reviewer: We probably want to link out to Ray Tune user guide instead?

.. _air-tuner:

Configuring Hyperparameter Tuning
=================================

The Ray AIR :class:`Tuner <ray.tune.Tuner>` is the recommended way to tune hyperparameters in Ray AIR.

.. image:: images/tuner.svg
    :align: center

> `Tuner` provides an interface that works with AIR Trainers to perform distributed

Reviewer 2: A `Tuner`

> hyperparameter tuning. `Tuner` is part of Ray Tune and provides

Reviewer 2: As part of Ray Tune, the `Tuner` provides ...

a variety of state-of-the-art hyperparameter tuning algorithms for optimizing model
performance.

> This page will cover basic usage examples of a Tuner. If you are interested in

Reviewer 2: Not a fan of "This page will..."

I would rephrase as: 

"What follows next is basic coverage of what a Tuner is and how you can use it for basic examples.

reading more, please take a look at the :ref:`Ray Tune documentation <tune-main>`.

Key Concepts
------------

> There are a couple key concepts for using a Tuner:

Reviewer 2: A think there is a list of items here not a couple. If you going to enumerate key concepts perhaps start with this:

There are a number of key concepts that dictate proper use of a Tuner:

> * First, you need to specify the hyperparameters you want to tune in a `search space`.

Reviewer 2: Remove first. Not needed.

 * A specific set of hyperparameters in a `search space.`
 * A `search algorithm` to ...
 * A set of arguments (`search space`, `search algorithm`, `scheduler`, and `Trainer`) passed to the `Tuner`
 *  Each individual hyperparater evalution run is called a `trial`
 * The `Tuner` returns its results in a `ResultGrid` object.

* You select a `search algorithm` to effectively optimize your parameters and optionally use a
  `scheduler` to stop searches early and speed up your experiments.
* Pass in the above objects with a `Trainer` into a `Tuner`,
  which runs the hyperparameter tuning workload by evaluating multiple hyperparameters in parallel.
* Each individual hyperparameter evaluation run is called a `trial`.
* The results are returned in a `ResultGrid`.

.. note::
   Tuners can also be used to launch hyperparameter tuning without using Ray AIR Trainers. See
   :ref:`the Ray Tune documentation <tune-main>` for more guides and examples.

Basic usage
-----------

Below, we demonstrate how you can use a Trainer object with a Tuner.

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __basic_start__
    :end-before: __basic_end__

> How do I configure a search space?

Reviewer 2: How to configure a search space?

I would leave out first person I here.

----------------------------------

A `Tuner` takes in a `param_space` argument where you can define the search space
from which hyperparameter configurations will be sampled.

Depending on the model and dataset, you may want to tune:

- The training batch size
- The learning rate for SGD-based training (e.g. image classification)

> - The maximum depth for tree-based models (e.g. XGBoost)

Reviewer 2: e.g.,

The following shows some example code on how to specify the ``param_space``.

.. tabbed:: XGBoost

    .. literalinclude:: doc_code/tuner.py
        :language: python
        :start-after: __xgboost_start__
        :end-before: __xgboost_end__

.. tabbed:: Pytorch

    .. literalinclude:: doc_code/tuner.py
        :language: python
        :start-after: __torch_start__
        :end-before: __torch_end__

Read more about :ref:`Tune search spaces here <tune-search-space-tutorial>`.

You can use a Tuner to tune most arguments and configurations in Ray AIR, including but
not limited to:

- Ray Datasets
- Preprocessors
- Scaling configurations
- and other hyperparameters.

There are a couple gotchas about parameter specification when using Tuners with Trainers:

> - By default, configuration dictionaries and config objects will be deep-merged

Reviewer 2: deep-merged.

> - Parameters that are duplicated in the Trainer and Tuner will be overwritten by the Tuner ``param_space``

Reviewer 2: ``param-space``.

- **Exception:** all arguments of the :class:`RunConfig <ray.air.config.RunConfig>` and :class:`TuneConfig <ray.tune.tune_config.TuneConfig>` are inherently un-tunable.

> How to configure a Tuner

Reviewer 2: How to configure a Tuner?

------------------------

There are two main configuration objects that can be passed into a Tuner: the :class:`TuneConfig <ray.tune.tune_config.TuneConfig>` and the :class:`RunConfig <ray.air.config.RunConfig>`.

The :class:`TuneConfig <ray.tune.tune_config.TuneConfig>` contains tuning specific settings, including:

- the tuning algorithm to use
- the metric and mode to rank results
- the amount of parallelism to use

Here are some common configurations for `TuneConfig`:

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __tune_config_start__
    :end-before: __tune_config_end__

See the :class:`TuneConfig API reference <ray.tune.tune_config.TuneConfig>` for more details.

The :class:`RunConfig <ray.air.config.RunConfig>` contains configurations that are more generic than tuning specific settings.
This may include:

- failure/retry configurations
- verbosity levels
- the name of the experiment
- the logging directory
- checkpoint configurations
- custom callbacks
- integration with cloud storage

> The following we showcase some common configuration of :class:`RunConfig <ray.air.config.RunConfig>`.

Reviewer 2: In the following section, we cover and showcase some common ...

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __run_config_start__
    :end-before: __run_config_end__

See the :class:`RunConfig API reference <ray.air.config.RunConfig>` for more details.

> How do I specify parallelism?

Reviewer 2: How to configure parallelism?

-----------------------------

You can specify parallelism via the :class:`TuneConfig <ray.tune.tune_config.TuneConfig>` by setting the following flags:

- `num_samples` which specifies the number of trials to run in total
- `max_concurrent_trials` which specifies the max number of trials to run concurrently

Note that actual parallelism can be less than `max_concurrent_trials` and will be determined by how many trials
can fit in the cluster at once (i.e., if you have a trial that requires 16 GPUs, your cluster has 32 GPUs,
and `max_concurrent_trials=10`, the `Tuner` can only run 2 trials concurrently).

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __tune_parallelism_start__
    :end-before: __tune_parallelism_end__

Read more about this in :ref:`tune-parallelism` section.

> How do I specify an optimization algorithm?

Reviewer 2: How to specify an optimization algorithm?

-------------------------------------------

You can specify your hyperparameter optimization method via the :class:`TuneConfig <ray.tune.tune_config.TuneConfig>` by setting the following flags:

- `search_alg` which provides an optimizer for selecting the optimal hyperparameters
- `scheduler` which provides a scheduling/resource allocation algorithm for accelerating the search process

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __tune_optimization_start__
    :end-before: __tune_optimization_end__

Read more about this in the :ref:`Search Algorithm <search-alg-ref>` and :ref:`Scheduler <schedulers-ref>` section.

How to analyze results?
-----------------------

> ``Tuner.fit()`` generates a ResultGrid object. This object contains metrics, results, and checkpoints

Reviewer 2: `ResultGrid`

> of each trial. Below is a simple usage example:

Reviewer 2: Below is a simple example:

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __result_grid_inspection_start__
    :end-before: __result_grid_inspection_end__

Advanced Tuning
---------------

> Tuners also offer the possibility to apply different data preprocessing steps, as shown in the following snippet.

Reviewer 2: should this be the "capability" instead of "possibility"?

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __tune_preprocess_start__
    :end-before: __tune_preprocess_end__

> You can also sample different train/validation datasets:

Reviewer 2: Additionally, you sample different ...

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __tune_dataset_start__
    :end-before: __tune_dataset_end__

Restoring and resuming
----------------------
A Tuner regularly saves its state, so that a tuning run can be resumed after being interrupted.

Additionally, if trials fail during a tuning run, they can be retried - either from scratch or
from the latest available checkpoint.

> To restore the Tuner state, you just pass the path to the experiment directory to ``Tuner.restore()``.

Reviewer 2: To restore the Tuner state, pass the path to the experiment directory as an argument to ``Tuner.restore(...)``

> This path can be obtained e.g. from the output of a tuning run (it's called "Result logdir"). If you

Reviewer 2: This path is obtained from the output of a tuning run, namely "Result logdir." However, if you specify a ``name`` in the :class:`RunConfig <ray.air.config.RunConfig>`, then it is located under ``~/ray_results/<name>``.

.. _air-tuner:

Hyperparameter Tuning
=====================

> Tuner API is the recommended way of launching hyperparameter tuning jobs with Ray Tune.

Reviewer: Add "what is a tuner and what does it do"?

    :start-after: __tune_preprocess_start__
    :end-before: __tune_preprocess_end__

Tuner API allows you to even tune the train/validation datasets! Examples coming soon!

In general, all the arguments accepted by your :ref:`Trainer <air-trainer-ref>` can be tunable.

Specify TuneConfig
------------------
This config contains tuning specific settings, including the tuning algorithm to use, the metric and mode to rank results etc.
See :ref:`Tuner <air-tuner-ref>` for details.

Specify RunConfig
-----------------
This config contains framework's runtime configurations that are more generic than tuning specific settings.
This may include failure/retry configurations, verbosity levels, the name of the experiment, its logging directory,
checkpoint configuration as well as its syncing configuration.

Putting everything together
---------------------------
We can now construct a Tuner and call ``Tuner.fit`` on it!

.. literalinclude:: doc_code/tuner.py
    :language: python
    :start-after: __tuner_start__
    :end-before: __tuner_end__

.. note::
    ``num_samples = 2`` here will be applied to the whole suite of grid search. In other words,
    we will generate 4 trials.

> For a more end-to-end example, checkout "analyze_tuning_results" example.

Author: Should we demote this to just an example rather than user guide?
Also how do I link into it?

Reviewer: link via `:ref:`
